### DISCUSSION TITLE
Position of allies in POV demos

### DISCUSSION THREAD

**Discussion by julian1198** [saul/demofile#304] [Questions] [answered]
I try to get the position of every player at each tick of a POV demo. My Position and the position of the opponents is changing, but the position of my allies stays the same.

    demoFile.on('tickend', e =>{
      if(e % 10000 !== 0) return;
      demoFile.players.forEach(element => {
          console.log(element.name);
          console.log(element.position);
      });
    });`

This is the code displays:

```
legendaryOverBot
{ x: 614.8709716796875, y: -11234.931640625, z: 0.03125 }
Ivko Sraszypko
{ x: -826.78466796875, y: -7.237809658050537, z: -104.42808532714844 }
...
legendaryOverBot
{ x: 614.8709716796875, y: -11234.931640625, z: 0.03125 }
Ivko Sraszypko
{ x: -1156.0220947265625, y: -773.5694580078125, z: -167.96875 }
```

As you can see the position of the ally has not changed after 10000 ticks and for the other allies it is the same. Is there a way to get the right position of the players?

**Comment by saul** [comment]
Hi @julian1198 - please can you attach the demo file if possible? I imagine the reason why the position isn't updated is because the allies in question are outside of the PVS (and so aren't networked to the recording player). You can check this with `element.isDormant`:

```js
demoFile.on('tickend', e =>{
  if(e % 10000 !== 0) return;
  demoFile.players.forEach(element => {
      console.log(element.name);
      if (element.isDormant)
        console.log('> dormant');
      else
        console.log(element.position);
  });
});
```

**Reply by julian1198** [reply]
I have added the demo file to this repository: [https://github.com/julian1198/demofile](https://github.com/julian1198/demofile)
I also tested if the allies are dormant. This was in none of the tested ticks the case. Only the opponents were sometimes dormant.

**Comment by saul** [comment, marked as answer]
This was a bug - apologies for taking so long to fix this. Your code should now work - I've just released v2.7.0 with the fix.

I've updated your code below to use the `ProcessSpottedEntityUpdate` user message. This allows you to get the position of enemies that have been spotted, but are outside of the recording player's PVS.

```ts
const spottedEntities = new Map<number, Vector>();
demoFile.userMessages.on("ProcessSpottedEntityUpdate", e => {
  if (e.newUpdate) spottedEntities.clear();
  for (const update of e.entityUpdates) {
    spottedEntities.set(update.entityIdx, {
      x: update.originX * 4,
      y: update.originY * 4,
      z: update.originZ * 4
    });
  }
});

demoFile.on("tickend", tick => {
  if (tick % 10000 !== 0) return;

  console.log("Tick", tick);

  demoFile.players.forEach(entity => {
    console.log(
      entity.clientSlot,
      "",
      entity.name,
      entity.isDormant ? "💤" : "",
      entity.isRecordingDemo ? "📹" : ""
    );

    if (entity.isDormant) {
      const pos = spottedEntities.get(entity.index);
      console.log("  approx", pos ? JSON.stringify(pos) : "<not spotted>");
    } else {
      console.log("  ", JSON.stringify(entity.position));
    }
  });

  console.log(`\n---\n`);
});
```
